x-pack/plugins/ml/public/application/explorer/swimlane_annotation_container.tsx

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { FC, useEffect } from 'react';
+import d3 from 'd3';
+import { scaleTime } from 'd3-scale';
+import { i18n } from '@kbn/i18n';
+import { formatHumanReadableDateTimeSeconds } from '../../../common/util/date_utils';
+import { AnnotationsTable } from '../../../common/types/annotations';
+import { ChartTooltipService } from '../components/chart_tooltip';
+
+const ANNOTATION_HEIGHT = 12;
+export const Y_AXIS_LABEL_WIDTH = 170;
+export const Y_AXIS_LABEL_PADDING = 8;
+export const Y_AXIS_LABEL_FONT_COLOR = '#6a717d';
+
+interface SwimlaneAnnotationContainerProps {
+  chartWidth: number;
+  domain: {
+    min: number;
+    max: number;
+  };
+  annotationsData?: AnnotationsTable['annotationsData'];
+  tooltipService: ChartTooltipService;
+}
+
+export const SwimlaneAnnotationContainer: FC<SwimlaneAnnotationContainerProps> = ({
+  chartWidth,
+  domain,
+  annotationsData,
+  tooltipService,
+}) => {
+  const canvasRef = React.useRef<HTMLDivElement | null>(null);
+
+  useEffect(() => {
+    if (canvasRef.current !== null && Array.isArray(annotationsData)) {
+      const chartElement = d3.select(canvasRef.current);
+      chartElement.selectAll('*').remove();
+
+      const startingXPos = Y_AXIS_LABEL_WIDTH + 2 * Y_AXIS_LABEL_PADDING;
+      const endingXPos = startingXPos + chartWidth - Y_AXIS_LABEL_PADDING;
+
+      const svg = chartElement
+        .append('svg')
+        .attr('width', '100%')
+        .attr('height', ANNOTATION_HEIGHT);
+
+      const xScale = scaleTime().domain([domain.min, domain.max]).range([startingXPos, endingXPos]);
+
+      // Add Annotation y axis label
+      svg
+        .append('text')
+        .attr('text-anchor', 'end')
+        .attr('class', 'swimlaneAnnotationLabel')
+        .text(
+          i18n.translate('xpack.ml.explorer.swimlaneAnnotationLabel', {
+            defaultMessage: 'Annotations',
+          })
+        )
+        .attr('x', Y_AXIS_LABEL_WIDTH + Y_AXIS_LABEL_PADDING)
+        .attr('y', ANNOTATION_HEIGHT)
+        .style('fill', Y_AXIS_LABEL_FONT_COLOR)
+        .style('font-size', '12px');
+
+      // Add border
+      svg
+        .append('rect')
+        .attr('x', startingXPos)
+        .attr('y', 0)
+        .attr('height', ANNOTATION_HEIGHT)
+        .attr('width', endingXPos - startingXPos)
+        .style('stroke', '#cccccc')
+        .style('fill', 'none')
+        .style('stroke-width', 1);
+
+      // Add annotation marker
+      annotationsData.forEach((d) => {
+        const annotationWidth = d.end_timestamp ? xScale(d.end_timestamp) - xScale(d.timestamp) : 0;
+        svg
+          .append('rect')
+          .classed('mlAnnotationRect', true)
+          .attr('x', xScale(d.timestamp))
```

Comment:
Looks like the x position and width of the annotation need to account for the time range of the overall swim lane:

![image](https://user-images.githubusercontent.com/7405507/115222649-dc1fa500-a102-11eb-8de9-309c60cdb797.png)

![image](https://user-images.githubusercontent.com/7405507/115222765-fc4f6400-a102-11eb-8162-7e4b1aa40db5.png)

Reply:
Fixed here [`985fced` (#97202)](https://github.com/elastic/kibana/pull/97202/commits/985fcede29b89b1cb306c3feb473f309b1b5777d)

Reply:
I don't see the `model_snapshot_stored` annotation at the end of the job any more. I do see it in the Single Metric Viewer though. Has one of the changes hidden it?

![image](https://user-images.githubusercontent.com/7405507/115431359-1bcab780-a1fd-11eb-90ae-03e38082ba2b.png)